src/Tuura/Plato/Translation.hs

```diff
@@ -19,36 +25,50 @@ instance Ord Signal
+        InvariantInitialState as ->
+            "The initial state is an invariant state.\nSignals with erroneous initial states: \n" ++ unlines (signs as) ++ "\n"
+    where
+        signs = map show
+
+  --   un ++ ic ++ ud ++ iv
+  -- where
+  --   un = if (unused  /= []) then ("The following signals are not declared as input, output or internal: \n" ++ unlines (map show unused) ++ "\n") else ""
+  --   ic = if (unused  /= []) then ("The following signals have inconsistent inital states: \n" ++ unlines (map show incons) ++ "\n") else ""
+  --   ud = if (undef   /= []) then ("The following signals have undefined initial states: \n" ++ unlines (map show undef) ++ "\n") else ""
+  --   iv = if (invInit /= []) then ("The initial state is an invariant state.\nSignals with erroneous initial states: \n" ++ unlines (map show invInit) ++ "\n") else ""
 
 validate :: Ord a => [a] -> CircuitConcept a -> ValidationResult a
 validate signs circuit
     | unused ++ inconsistent ++ undef ++ invInit == [] = Valid
-    | otherwise                             = Invalid unused inconsistent undef invInit
+    | otherwise = Invalid (
+        (if unused       /= [] then [UnusedSignal unused]                   else []) ++
+        (if inconsistent /= [] then [InconsistentInitialState inconsistent] else []) ++
+        (if undef        /= [] then [UndefinedInitialState undef]           else []) ++
+        (if invInit      /= [] then [InvariantInitialState invInit]         else []))
   where
     unused       = filter ((==Unused) . interface circuit) signs
     inconsistent = filter ((==Inconsistent) . initial circuit) signs
     undef        = filter ((==Undefined) . initial circuit) signs
-    invInit      = checkInitialStates signs circuit
+    invInit      = concatMap (\i -> checkInitialStates signs i (initial circuit)) (invariant circuit)
 
-checkInitialStates :: Ord a => [a] -> CircuitConcept a -> [a]
-checkInitialStates signs circuit = nubOrd (concatMap (\inv -> if (checkForInvariant inv initialStates) then (invariantError inv) else []) (invariant circuit))
+checkInitialStates :: Ord a => [a] -> Invariant (Transition a) -> (a -> InitialValue) -> [a]
+checkInitialStates signs (NeverAll es) initials = nubOrd (if (checkForInvariant es initialStates) then (invariantError es) else [])
   where
-    initialStates = map (\s -> Transition { signal = s , newValue = getDefined $ initial circuit s }) signs
+    initialStates = map (\s -> Transition { signal = s , newValue = getDefined $ initials s }) signs
 
 invariantError :: [Transition a] -> [a]
-invariantError = map (signal)
+invariantError = map signal
 
 checkForInvariant :: Eq a => [Transition a] -> [Transition a] -> Bool
-checkForInvariant invs initialStates
-    | invs == [] = True
-    | otherwise  = ((head invs) `elem` initialStates) && (checkForInvariant (tail invs) initialStates)
+checkForInvariant invs initialStates = all (`elem` initialStates) invs
 
```

Comment:
I agree, what a nice simple function :) I might remove the function actually, and just have:
`all elem initialStates` instead.